oda_api/token.py

```diff
@@ -62,57 +69,125 @@ def decode_oda_token(token: str, secret_key=None, allow_invalid=False) -> dict:
 
     raise RuntimeError()
 
+
 def decode_oauth2_token(token: str):
     # usually comes in cookies['_oauth2_proxy']
     return json.loads(base64.b64decode(token.split(".")[0]+"=").decode())
 
-#TODO: move to dynaconf
-def discover_token(
+
+def get_token_roles(decoded_token):
+    # extract role(s)
+    roles = None
+    if isinstance(decoded_token['roles'], str):
+        roles = decoded_token['roles'].split(',') if 'roles' in decoded_token else []
+        roles[:] = [r.strip() for r in roles]
+    elif isinstance(decoded_token['roles'], list):
+        roles = decoded_token['roles'] if 'roles' in decoded_token else []
+        roles[:] = [r.strip() for r in roles]
+    return roles
+
+
+def rewrite_token(new_token,
+                  token_write_method: TokenAccessMethods = None
+                  ):
+    current_token, discover_method = discover_token_and_method(allow_invalid=True)
+    current_decoded_token = decode_oda_token(current_token, allow_invalid=True)
+    new_decoded_token = decode_oda_token(new_token, allow_invalid=True)
+
+    current_decoded_token_expires_in_s = current_decoded_token['exp'] - time.time()
+    new_decoded_token_expires_in_s = new_decoded_token['exp'] - time.time()
+
+    if new_decoded_token_expires_in_s < current_decoded_token_expires_in_s:
+        logger.info("the new token has will expire before the current one")
+
+    current_decoded_token_roles = get_token_roles(current_decoded_token)
+    new_decoded_token_roles = get_token_roles(new_decoded_token)
+    if set(current_decoded_token_roles) - set(new_decoded_token_roles) != set():
+        logger.info("The new token has a different set of roles than the current one:\n"
```

Comment:
I'd say if new roles include all old roles - logger.warning("new token has more roles than the existing one: xxxx"). If new roles are fewer - a warning and return, unless "force=True".